mmverify.py

```diff
@@ -1,369 +1,588 @@
+by verifying compressed proofs without converting them to normal proof format;
+added type hints
+"""
 
 import sys
 import itertools
-import collections
-import os.path
-from optparse import OptionParser
-
-verbosity = 1
-
-class MMError(Exception): pass
-class MMKeyError(MMError, KeyError): pass
-
-def vprint(vlevel, *args):
-    if verbosity >= vlevel: print(*args, file=sys.stderr)
-
-class toks:
-    def __init__(self, lines):
-        self.lines_buf = [lines]
-        self.tokbuf = []
-        self.imported_files = set()
-
-    def read(self):
-        while self.tokbuf == []:
+import pathlib
+import argparse
+import typing
+import io
+
+Label = str
+Var = str
+Const = str
+Steptyp = str  # can actually be one of '$e', '$f', '$a', '$p'
+StringOption = typing.Optional[str]
+Symbol = typing.Union[Var, Const]
+Stmt = list[Symbol]
+Ehyp = Stmt
+Fhyp = tuple[Var, Const]
+Dv = tuple[Var, Var]
+Assertion = tuple[set[Dv], list[Fhyp], list[Ehyp], Stmt]
+FullStmt = tuple[Steptyp, typing.Union[Stmt, Assertion]]
+# Actually, the second component of a FullStmt is a Stmt when its first
+# component is '$e' or '$f' and an Assertion if its first component is '$a' or
+# '$p', but this is a bit cumbersome to build it into the typing system.
+# This explains the errors when static type checking (e.g., mypy): an
+# if-statement determines in which case we are, but this is invisible to the
+# type checker.
+
+
+class MMError(Exception):
+    """Class of Metamath errors."""
+    pass
+
+
+class MMKeyError(MMError, KeyError):
+    """Class of Metamath key errors."""
+    pass
+
+
+def vprint(vlevel: int, *arguments: typing.Any) -> None:
+    """Print log message if verbosity level is higher than the argument."""
+    if verbosity >= vlevel:
+        print(*arguments, file=logfile)
+
+
+class Toks:
+    """Class of sets of tokens from which functions read as in an input
+    stream.
+    """
+
+    def __init__(self, file: io.TextIOWrapper) -> None:
+        """Construct a 'Toks' from the given file: initialize a line buffer
+        containing the lines of the file, and initialize a set of imported
+        files to a singleton containing that file, so as to avoid multiple
+        imports.
+        """
+        self.lines_buf = [file]
+        self.tokbuf: list[str] = []
+        self.imported_files = set({pathlib.Path(file.name).resolve()})
+
+    def read(self) -> StringOption:
+        """Read the next token in the token buffer, or if it is empty, split
+        the next line into tokens and read from it."""
+        while not self.tokbuf:
             line = self.lines_buf[-1].readline()
             if not line:
                 self.lines_buf.pop().close()
-                if not self.lines_buf: return None
+                if not self.lines_buf:
+                    return None
             else:
                 self.tokbuf = line.split()
                 self.tokbuf.reverse()
         return self.tokbuf.pop()
```

Comment:
Obviously this is not from your code, but the `reverse` here is a bit surprising at first sight.
I assume it's more efficient to `pop` the last element than the first one. 
A `queue` might be even more efficient, but there might be overhead to build it.

Reply:
As I understand, there is no "stack" in Python, but lists can be used as stacks and pop/appends happen at the end of the list (https://docs.python.org/3/tutorial/datastructures.html?highlight=stack#using-lists-as-stacks), hence the need to reverse it. It may not be worth importing Queue for that. Or maybe you had an intermediate solution in mind ?

Reply:
Looking at it again, if we're looking for performance, I guess the most efficient would probably to use an [iterator](https://docs.python.org/3/library/functions.html#next). That would be something like:
```python
   self.tokbuf = line.split()
   return self.tokbuf.next()
```
and testing for the [StopIteration](https://docs.python.org/3/library/exceptions.html#StopIteration) exception.
This shall avoid reverting each line, I suppose there could be another small performance gain.

But no need to do this in this PR, though.